tevfik_aksoy/graph.cpp

```diff
@@ -0,0 +1,177 @@
+  return edge_ids_;
+}
+
+std::string color_to_string(const Edge::Color& color) {
+  switch (color) {
+    case Edge::Color::Gray:
+      return "gray";
+    case Edge::Color::Green:
+      return "green";
+    case Edge::Color::Blue:
+      return "blue";
+    case Edge::Color::Yellow:
+      return "yellow";
+    case Edge::Color::Red:
+      return "red";
+    case Edge::Color::Unknown:
+      return "unknown";
+  }
+}
+
+std::string Edge::to_JSON() const {
+  std::string json_string;
+  json_string += "\t{ \"id\": " + std::to_string(id) + ", \"vertex_ids\": [" +
+                 std::to_string(source) + ", " + std::to_string(destination) +
+                 R"(], "color": ")" + color_to_string(color) + "\" }";
+  return json_string;
+}
+
+std::string Vertex::to_JSON() const {
+  std::string json_string;
+  json_string += "\t{ \"id\": " + std::to_string(id) + ", \"edge_ids\": [";
+  for (int i = 0; i < edge_ids_.size(); i++) {
+    json_string += std::to_string(edge_ids_[i]);
+    if (i + 1 != edge_ids_.size())
+      json_string += ", ";
+  }
+  json_string += "], \"depth\": " + std::to_string(depth) + "}";
+  return json_string;
+}
+
+bool Graph::does_vertex_exist(const VertexId& id,
+                              const std::vector<Vertex>& vertices) const {
+  for (const auto& vertex : vertices)
+    if (vertex.id == id) {
+      return true;
+    }
+  return false;
+}
+
+VertexId Graph::insert_vertex() {
+  const auto id = get_new_vertex_id_();
+  vertices_.emplace_back(id);
+  if (id == 0) {
+    VertexDepth depth = 0;
+    depth_map_.emplace_back();
+    depth_map_[depth].push_back(id);
+  }
+  return id;
+}
+
+Edge::Color Graph::calculate_color_for_edge_(const Vertex& source,
+                                             const Vertex& destination) const {
+  if (source.get_edge_ids().empty() || destination.get_edge_ids().empty())
+    return Edge::Color::Gray;
+  if (source.id == destination.id)
+    return Edge::Color::Green;
+  if (source.depth == destination.depth) {
+    for (int i = 0; i < depth_map_[source.depth].size() - 1; i++) {
+      auto first = depth_map_[source.depth][i];
+      auto second = depth_map_[source.depth][i + 1];
+      if ((source.id == first && destination.id == second) ||
+          (destination.id == first && source.id == second))
+        return Edge::Color::Blue;
+    }
+  }
+  if (source.depth == destination.depth - 1)
+    return Edge::Color::Yellow;
+  if (source.depth == destination.depth - 2)
+    return Edge::Color::Red;
+
+  // unknown нужно для return val of calculate_edge_color
+  return Edge::Color::Unknown;
+}
+
+void Graph::insert_edge(const VertexId& source_id,
+                        const VertexId& destination_id) {
+  assert(!are_vertices_connected(source_id, destination_id) &&
+         "Vertices are already connected!");
+  const auto color = calculate_color_for_edge_(vertices_[source_id],
+                                               vertices_[destination_id]);

```

Comment:
У тебя нет прямой связи между индексом и `id`, ты должен непосредственно искать вершину в коллекции по `id`, можно эту логику вынести в функцию
```suggestion
  const auto& source_vertex = get_vertex(source_id);
  const auto& destination_vertex = get_vertex(destination_id);
  const auto color = calculate_color_for_edge_(source_vertex,
                                               destination_vertex);
```
Ты мог бы использовать `vertices_[vertex_id]`, если бы у тебя вершины были `unordered_map`
```cpp
std::unordered_map<VertexId, Vertex> vertices_;
```
Тогда да, ты мог бы получать вершину сразу по её `id`

Reply:
Сделал get_vertex